File: src/c/librosie/new_ffi_test.rb
# gem install ffi

require 'ffi'
require 'json'

$: << "."                       # temporary: add "." to the library search path
require 'rosie'

def print_string_array(retval)
  print "number of strings returned from api call is: ", retval[:n], "\n"
  ptr_array = FFI::Pointer.new(FFI::Pointer, retval[:ptr]).read_array_of_pointer(retval[:n])
  for i in 0..(retval[:n]-1) do
    str = Rosie::CString.new ptr_array[i]
    print "  [", i, "] len=", str[:len], ", ptr=", str[:ptr].read_string_length(str[:len]), "\n"
  end
end

def string_array_from_CStringArray(retval)
  ptr_array = FFI::Pointer.new(FFI::Pointer, retval[:ptr]).read_array_of_pointer(retval[:n])
  strings = []
  for i in 0..(retval[:n]-1) do
    cstr = Rosie::CString.new ptr_array[i]
    strings[i] = cstr[:ptr].read_string_length(cstr[:len])
  end
  return strings
end

messages = Rosie::CStringArray.new
engine = Rosie.initialize("/Users/jjennings/Work/Dev/rosie-pattern-language", messages)
print_string_array(messages)


config_string = Rosie.CString_from_string("{\"name\":\"Ruby engine\"}")
print "config_string string: ", config_string[:ptr].read_string_length(config_string[:len]), "\n"
print "config_string length: ", config_string[:len], "\n"
print "config_string struct size: ", Rosie::CString.size, "\n"

ignored = Rosie::CString_from_string("ignored")

config_js = "{\"expression\" : \"\\\"ign\\\"\", \"encode\" : \"json\"}"
config_string = Rosie::CString_from_string(config_js)

retval = Rosie.configure_engine(engine, config_string)
print_string_array(retval)
Rosie.free_stringArray(retval)

retval = Rosie.inspect_engine(engine)
print_string_array(retval)
Rosie.free_stringArray(retval)

retval = Rosie.rosie_api(engine, "match", ignored)
print_string_array(retval)
Rosie.free_stringArray(retval)

retval = Rosie.rosie_api(engine, "match", config_string)
print_string_array(retval)
Rosie.free_stringArray(retval)

config_string = Rosie::CString_from_string("This is NOT valid json")
retval = Rosie.configure_engine(engine, config_string.pointer)
print_string_array(retval)
Rosie.free_stringArray(retval)

retval = Rosie.rosie_api(engine, "load_manifest", ignored) # should fail
print_string_array(retval)
Rosie.free_stringArray(retval)

retval = Rosie.configure_engine(engine, config_string)
print_string_array(retval)
Rosie.free_stringArray(retval)

# Loop test prep

test = Rosie::CString_from_string("$sys/MANIFEST")
print "TEST: len=", test[:len], "\n"
print "TEST: string=", test[:ptr].read_string, "\n"

retval = Rosie.rosie_api(engine, "load_manifest", test)
print_string_array(retval)
Rosie.free_stringArray(retval)

config_js = "{\"expression\" : \"[:digit:]+\", \"encode\" : \"json\"}"
print "config_js is: ", config_js, "\n"

config_string = Rosie::CString_from_string(config_js)
retval = Rosie.rosie_api(engine, "configure_engine", config_string)
print_string_array(retval)
Rosie.free_stringArray(retval)

# Loop test

foo = Rosie::CString_from_string("1239999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999")
retval = Rosie.rosie_api engine, "match", foo
retval_SAVE = retval

call_rosie = true

print "Looping..."
M = 1000000
for i in 0..5*M do
#for i in 0..3 do
  if call_rosie then
    retval = Rosie.match engine, foo
  else
    retval = Rosie::CStringArray.new; retval[:n] = retval_SAVE[:n]; retval[:ptr] = retval_SAVE[:ptr]
  end
  strings = string_array_from_CStringArray(retval)
  code = strings[0]
  if code != "true" then
    print "Error code returned from match api"
  end
  json_string = strings[1]
  if call_rosie then
    Rosie.free_stringArray(retval)
  end
  # if code=="true" then
  #  print "Successful call to match\n"
  # else
  #   print "Call to match FAILED\n"
  # end
  # print json_string, "\n"
  obj_to_return_to_caller = JSON.parse(json_string)
end

print " done.\n"
Rosie.finalize(engine)

File: src/c/librosie/new_ffi_test.js
//  -*- Mode: Javascript; -*-                                      
// 
//  new_ffi_test.js
// 
//  © Copyright IBM Corporation 2016.
//  LICENSE: MIT License (https://opensource.org/licenses/mit-license.html)
//  AUTHOR: Jamie A. Jennings

//
// npm install ffi
//

var debug = require("debug");

var ffi = require("ffi");
var ref = require('ref');
var Struct = require('ref-struct');
var ArrayType = require('ref-array')

var DynamicLibrary = require('./node_modules/ffi/lib/dynamic_library')
  , ForeignFunction = require('./node_modules/ffi/lib/foreign_function')


var MyCString = Struct({
    'len': 'uint32',
    'ptr': ref.refType('CString')
});

var MyCStringArray = Struct({
    'n': 'uint32',
    'ptr': ref.refType(ref.refType(MyCString))
});

var MyCStringPtr = ref.refType(MyCString)
var MyCStringArrayPtr = ref.refType(MyCStringArray)

var RTLD_NOW = ffi.DynamicLibrary.FLAGS.RTLD_NOW;
var RTLD_GLOBAL = ffi.DynamicLibrary.FLAGS.RTLD_GLOBAL;
var mode = RTLD_NOW | RTLD_GLOBAL;

var RosieLib = new DynamicLibrary('librosie.so' || null, mode);

var funcs = {'initialize': [ 'pointer', ['string', MyCStringArrayPtr]],
	     'inspect_engine': [ MyCStringArray, ['pointer'] ],
	     'configure_engine': [ MyCStringArray, ['pointer', MyCStringPtr] ],
	     'inspect_engine': [ MyCStringArray, ['pointer'] ],
	     'rosie_api': [ MyCStringArray, ['pointer', 'string', MyCStringPtr] ],
	     'finalize': [ 'void', ['pointer']],
	     'free_stringArray': [ 'void', [MyCStringArray] ]
	    }

var Rosie = new ffi.Library;
Object.keys(funcs || {}).forEach(function (func) {
    debug('defining function', func)

    var fptr = RosieLib.get(func)
      , info = funcs[func]

    if (fptr.isNull()) {
      throw new Error('Library: "' + Libfile
        + '" returned NULL function pointer for "' + func + '"')
    }

    var resultType = info[0]
      , paramTypes = info[1]
      , fopts = info[2]
      , abi = fopts && fopts.abi
      , async = fopts && fopts.async
      , varargs = fopts && fopts.varargs

    if (varargs) {
      Rosie[func] = VariadicForeignFunction(fptr, resultType, paramTypes, abi)
    } else {
      var ff = ForeignFunction(fptr, resultType, paramTypes, abi)
      Rosie[func] = async ? ff.async : ff
    }
  })


var buf = ref.allocCString('hello world');
buf[3] = 0
console.log(buf.length, buf.toString());

var str = new MyCString
str.ptr = ref.allocCString("Hi!").ref()
str.len = str.ptr.deref().length
console.log("str.len =", str.len, " and str.ptr.length =", str.ptr.length)
console.log("str len=", str.len, " and ptr=", str.ptr.deref().toString("utf8",0,str.len))

console.log(buf.length, buf.toString());
var xx = buf.ref()
var x = xx.deref().reinterpret(Buffer.byteLength(buf.toString()))
x = x.ref()
console.log(buf.length, x.deref().toString().length, Buffer.byteLength(buf.toString()))
console.log("x is:", x, " x.length=", x.length, " and ptr=", x.deref().toString("utf8", 0, buf.length))
console.log("x.deref() is:", x.deref())
x.deref()[1]=0
console.log("x is:", x, " x.length=", x.length, " and ptr=", x.deref().toString("utf8", 0, buf.length))
console.log("x.deref() is:", x.deref())

function new_CString(str) {
    var cstr = new MyCString
    cstr.ptr = ref.allocCString(str)
    cstr.len = str.length
    return cstr
}

function extract_string_from_CString_ptr(cstr_ptr) {
    var cstr = ref.alloc(MyCString)
    cstr = cstr_ptr.deref()
    var buf = cstr.ptr.reinterpret(cstr.len)
    return buf.toString()
}    

function extract_string_from_array(stringarray, index) {
    var n = stringarray.n
    var p = stringarray.ptr
    if (index < n) {
	var cstr_ptr = ref.get(p, (index*ref.sizeof.pointer), MyCStringPtr)
	var str = extract_string_from_CString_ptr(cstr_ptr)
	return str
    }
    else return null
}

function print_array(retval) {
    var n = retval.n
    for (var i=0; i<n; i++) {
	var str = extract_string_from_array(retval, i)
	console.log("print_array: [", i, "] length =", str.length, " and value =", str)
    }
}

messages = new MyCStringArray
console.log("About to initialize Rosie")
var engine = Rosie.initialize("/Users/jjennings/Work/Dev/rosie-pattern-language", messages.ref())
console.log("Return value from initialize: ", engine)

var config = new MyCString
var tbl = {'name': 'JS test engine', 'expression':'[:digit:]+', 'encode':false}
var tmp = JSON.stringify(tbl)
config.ptr = ref.allocCString(tmp)
config.len = Buffer.byteLength(tmp)
var buf = config.ptr.reinterpret(config.len)
//console.log(config.ptr.length, config.len, buf.toString("utf8", 0, config.len))

var i = ref.alloc(MyCStringArray)
i = Rosie.configure_engine(engine, config.ref())
console.log("Return value from configure_engine is: ")
var code = extract_string_from_array(i, 0)
print_array(i)

var ignored = new_CString("ignored")

i = Rosie.inspect_engine(engine)
console.log("Return value from inspect_engine is: ")
print_array(i)

// Loop prep

var manifest = new_CString("$sys/MANIFEST")
console.log("manifest: len =", manifest.len, "value =", manifest.ptr.toString())

var retval = Rosie.rosie_api(engine, "load_manifest", manifest.ref())
print_array(retval)
Rosie.free_stringArray(retval)

var config = new_CString("{\"expression\" : \"[:digit:]+\", \"encode\" : \"json\"}")

retval = Rosie.configure_engine(engine, config.ref())
print_array(retval)
Rosie.free_stringArray(retval)

var foo = new_CString("1239999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999")
var retval = Rosie.rosie_api(engine, "match", foo.ref())
retval_SAVE = retval
print_array(retval_SAVE)

// Loop

console.log("Looping...")
M = 1000000
call_rosie = true
for (var i=0; i<5*M; i++) {
//for (var i=0; i<5; i++) {
    if (call_rosie) retval = Rosie.rosie_api(engine, "match", foo.ref())
    else {
	retval = new MyCStringArray;
	retval.n = retval_SAVE.n;
	retval.ptr = retval_SAVE.ptr;
    }
    code = extract_string_from_array(retval, 0)
    if (code != "true") console.log("Error code returned from match api")
    json_string = extract_string_from_array(retval, 1)
    if (call_rosie) Rosie.free_stringArray(retval)
//    if (code=="true") console.log("Successful call to match")
//    else console.log("Call to match FAILED")
//    console.log(json_string)
    obj_to_return_to_caller = JSON.parse(json_string)
}
console.log("Done.\n")
